client/components/edit/StateModal.tsx

```diff
@@ -0,0 +1,188 @@
+    } = useFetchContext();
+    const [statusMessage, setStatusMessage] = useState<string>("");
+    const [includeChildren, setIncludeChildren] = useState<boolean>(false);
+    const [selectedValue, setSelectedValue] = useState<string>("Inactive");
+    const [childPidResponse, setChildPidResponse] = useState({ loading: true });
+    const loaded = Object.prototype.hasOwnProperty.call(objectDetailsStorage, stateModalActivePid);
+    const details = loaded ? objectDetailsStorage[stateModalActivePid] : {};
+    const childPageSize = 1000;
+    useEffect(() => {
+        async function loadChildren() {
+            setChildPidResponse({ loading: true });
+            setIncludeChildren(false);
+            const url = getObjectRecursiveChildPidsUrl(details.pid, 0, childPageSize);
+            const response = await fetchJSON(url);
+            setChildPidResponse(response);
+        }
+        setSelectedValue(details.state ?? "Inactive");
+        if (details.pid ?? false) {
+            loadChildren();
+        }
+    }, [details]);
+
+    const handleRadioChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+        setSelectedValue(event.target.value);
+    };
+
+    const handleChildCheckboxChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+        setIncludeChildren(event.target.checked);
+    };
+
+    const showSnackbarMessage = (message: string, severity: string) => {
+        setSnackbarState({
+            open: true,
+            message,
+            severity,
+        });
+    };
+
+    const updateStatus = async (pid: string): Promise<string> => {
+        setStatusMessage(`Saving status for ${pid}...`);
+        const target = getObjectStateUrl(pid);
+        const result = await fetchText(target, { method: "PUT", body: selectedValue });
+        if (result === "ok") {
+            // Clear and reload the cached object, since it has now changed!
+            removeFromObjectDetailsStorage(pid);
+        }
+        return result;
+    };
+
+    const saveChildPage = async (response): Promise<boolean> => {
+        for (let i = 0; i < response.docs.length; i++) {
+            const result = await updateStatus(response.docs[i].id);
+            if (result !== "ok") {
+                showSnackbarMessage(`Status failed to save; "${result}"`, "error");
+                toggleStateModal();
+                setStatusMessage("");
+                return false;
+            }
+        }
+        return true;
+    };
+
+    const saveChildren = async (): Promise<boolean> => {
+        const expectedTotal = childPidResponse.numFound;
+        let found = 0;
+        let nextResponse = childPidResponse;
+        while (found < expectedTotal) {
+            if (!(await saveChildPage(nextResponse))) {
+                return false;
+            }
+            found += nextResponse.docs.length;
+            if (found < expectedTotal) {
+                const url = getObjectRecursiveChildPidsUrl(details.pid, found, childPageSize);
```

Comment:
Is it possible to batch all the saves or do these operations in the backend?

Reply:
In theory, yes, and this might be a future enhancement. For now, I'm matching the functionality of the legacy system, which does the updates in-browser and provides real-time status updates. This is actually useful in many situations, since the user usually wants to be immediately informed when this task is completed. Some kind of asynchronous batch-editing tool will likely be needed in the future, though, and that could provide an alternative for cases where the user doesn't want to tie up their browser waiting for state changes to finish.